LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DeviceLoaderSynchronizer.cs

```diff
@@ -0,0 +1,330 @@
+                        this.AddToDeviceQueue(device, request);
+                        return;
+                    }
+                }
+
+                // not handled, raised failed event
+                var failedReason =
+                    this.loadingDevicesFailed ? LoRaDeviceRequestFailedReason.ApplicationError :
+                    this.destinationDictionary.Count > 0 ? LoRaDeviceRequestFailedReason.NotMatchingDeviceByMicCheck : LoRaDeviceRequestFailedReason.NotMatchingDeviceByDevAddr;
+
+                this.LogRequestFailed(request, failedReason);
+
+                request.NotifyFailed(null, failedReason);
+            }
+        }
+
+        private void AddToDeviceQueue(LoRaDevice device, LoRaRequest request)
+        {
+            if (device.IsOurDevice)
+            {
+                device.Queue(request);
+            }
+            else
+            {
+                Logger.Log(device.DevEUI, $"device is not our device, ignore message", LogLevel.Debug);
+                request.NotifyFailed(device, LoRaDeviceRequestFailedReason.BelongsToAnotherGateway);
+            }
+        }
+
+        void SetState(LoaderState newState)
+        {
+            if (this.state != newState)
+            {
+                Logger.Log(this.devAddr, $"Loader state changed {this.state} => {newState}", LogLevel.Debug);
+                this.state = newState;
+            }
+        }
+
+        private void LogRequestFailed(LoRaRequest request, LoRaDeviceRequestFailedReason failedReason)
+        {
+            var deviceId = ConversionHelper.ByteArrayToString(request.Payload.DevAddr);
+
+            switch (failedReason)
+            {
+                case LoRaDeviceRequestFailedReason.NotMatchingDeviceByMicCheck:
+                    Logger.Log(deviceId, $"with devAddr {ConversionHelper.ByteArrayToString(request.Payload.DevAddr)} check MIC failed", LogLevel.Debug);
+                    break;
+
+                case LoRaDeviceRequestFailedReason.NotMatchingDeviceByDevAddr:
+                    Logger.Log(deviceId, $"device is not our device, ignore message", LogLevel.Information);
+                    break;
+
+                case LoRaDeviceRequestFailedReason.ApplicationError:
+                    Logger.Log(deviceId, "problem resolving device", LogLevel.Information);
+                    break;
+            }
+        }
+
+        private async Task<LoRaDevice> InitializeDeviceAsync(LoRaDevice loRaDevice)
+        {
+            try
+            {
+                // Calling initialize async here to avoid making async calls in the concurrent dictionary
+                // Since only one device will be added, we guarantee that initialization only happens once
+                if (await loRaDevice.InitializeAsync())
+                {
+                    loRaDevice.IsOurDevice = string.IsNullOrEmpty(loRaDevice.GatewayID) || string.Equals(loRaDevice.GatewayID, this.configuration.GatewayID, StringComparison.InvariantCultureIgnoreCase);
```

Comment:
can be confusing for a newbie like me if IsOurDevice has multiple meanings. If there is no gateway id, it's not really 'our' device right? I thought it was our device if the gatewayid was explicitly set and matches.

Reply:
`IsOurDevice` here means whether or not we should handle the device messages. That is true if the device.GatewayID matches ours or the device does not have a gatewayID (in that case all gateways will process it's messages).

Could be changed to IsExternalDevice that would mean device.GatewayID != ourGatewayID. What do you think?

Reply:
IsExternalDevice would make it clearer to me. A comment on the property would also help. :)

Reply:
IsOurDevice means that it belongs to our IoT Hub instance (basically is in the DB behind IoT Hub

Reply:
With the changes introduced in 0.3-preview this flag indicates if the messages of a device should  be handled by the network module.

A device in the join process or a device that is assigned to a different gateway will have IsOurDevice to `false`. All others will have `true`

Reply:
Then we should rename it to something more meaningful IsDeviceLinkedToThisGateway